javascript/ecmascript-2015/function-improvements/arrow-functions-part-2.md

```diff
@@ -40,26 +40,51 @@ links:
 ---
 ## Content
 
-Arrow Functions(`() => {}`) differ from anonymous keyword functions (`function(){}`) by not changing the value of `this`.
+Arrow Functions(`() => {}`) differ from anonymous keyword functions (`function(){}`) by not changing the value of `this`. Before **ES6**, every time you would call a function it would define its own `this` value, depending on how the function was called.
+
+Let's take a look at how this `Counter()` constructor is being defined: 
 
 ```js
 window.counter = 0;
 
 function Counter() {
+  // The Counter() constructor defines `this`
+  // as an instance of itself.
   this.counter = 0;
+
   this.countUp = setInterval(function() {
+    // The countUp() function defines `this`
```

Comment:
So this snippet has a few important mistakes.

```js
var a = setInterval(...)
```

will immediately run the `setInterval` (and start the periodic invocations of the given function) and return an id to use for `clearInterval` (to stop the periodic invocations).

This means that in this example `countUp` and `countDown` are just numbers and could be removed entirely.

It also treats `this` as a static property (using words like "defined") but `this` entirely depends on how a function is called.

These invocations will all have a different `this` in `Counter`:

```js
Counter();
new Counter();
Counter.call('example')
```

Here's one way to improve the snippet. Feel free to reword this further:

```js
function Counter() {
  // We wrote `Counter` in such
  // a way that it expects `this`
  // to be an instance of itself.
  // In other words, we expect
  // `Counter` to be called using the
  // keyword `new`, which will initialize
  // `this` to be an instance object, and
  // we assign a `counter` variable to
  // that object
  this.counter = 0;

  setInterval(function () {
    // This function will have `this`
    // as the global object because setInterval
    // will call it like a regular function, i.e. fn()

    // this means that in here we increment a `counter`
    // on the global object, not the one on `Counter`
    this.counter++;
  }, 1000)

  setInterval(() => {
    // this arrow function will look up `this`
    // in the scope above, which is the 
    // scope of Counter, and use that.
    // In here, we update the `counter`
    // on whatever `this` points to 
    // within `Counter`, which in our 
    // case should be an instance of
    // `Counter`
  })
}

let count = new Counter();
```

Reply:
I should've paid more attention when writing this. I was basing my explanation of the MDN docs where they use a `Person()` constructor, but forgot we are using `Counter()`. I'll fix these right now.